singularity-eos/closure/mixed_cell_models.hpp

```diff
@@ -272,450 +129,636 @@ PORTABLE_INLINE_FUNCTION bool solve_Ax_b(Real *a, Real *b) {
-    b[2 * nmat - 1] = vsum - Volume;
-    for (int i = 0; i < 2 * nmat; ++i)
-      err += square(b[i]);
-    err *= 0.5;
-    if (!solve_Ax_b<2 * nmat>(A, b)) {
-      // do something to crash out?  Tell folks what happened?
-#ifndef KOKKOS_ENABLE_CUDA
-      printf("Crashing out on iteration %i\n", iter);
-#endif // KOKKOS_ENABLE_CUDA
-      break;
+    // Normalize vfrac
+    for (int m = 0; m < nmat; ++m) {
+      vfrac[m] *= vfrac_total / vsum;
+    }
+    // check to make sure the normalization didn't put us below rho_at_pmin
+    for (int m = 0; m < nmat; ++m) {
+      const Real rho_min = eos[m].RhoPmin(Tguess);
+      rho[m] = rhobar[m] / vfrac[m];
+      if (rho[m] < rho_min) {
+        printf("rho < rho_min in PTE initialization!  Solver may not converge.");
+      }
     }
-    // LINE SEARCH
 
-    // Now, get overall scaling limit
-    Real scale = 1.0;
+    for (int m = 0; m < nmat; m++) {
+      temp[m] = 1.0;
+      sie[m] = eos[m].InternalEnergyFromDensityTemperature(rho[m], Tguess);
+      if (eos[m].PreferredInput() ==
+          (thermalqs::density | thermalqs::specific_internal_energy)) {
+        press[m] = eos[m].PressureFromDensityInternalEnergy(rho[m], sie[m], Cache[m]) / u_total;
+      } else if (eos[m].PreferredInput() ==
+                (thermalqs::density | thermalqs::temperature)) {
+        press[m] = eos[m].PressureFromDensityTemperature(rho[m], Tguess, Cache[m]) / u_total;
+      }
+    }
+
+    for (int m = 0; m < nmat; ++m)
+      u[m] = sie[m] * rhobar[m] / u_total;
+  }
+
+  PORTABLE_INLINE_FUNCTION
+  Real ResidualNorm() const {
+    Real norm = 0.0;
+    for (int m = 0; m < neq; m++) norm += residual[m]*residual[m];
+    return 0.5*norm;
+  }
+
+  PORTABLE_FORCEINLINE_FUNCTION
+  int MatIndex(const int &i, const int &j) const {
+    return i*neq + j;
+  }
+
+  PORTABLE_INLINE_FUNCTION
+  void GetIdealPTE(Real &Pideal, Real &Tideal) const {
+    Real rhoBsum = 0.0;
+    Real Asum = 0.0;
+    for (int m = 0; m < nmat; m++) {
+      Asum += vfrac[m] * press[m] / temp[m];
+      rhoBsum += rho[m] * vfrac[m] * sie[m] / temp[m];
+    }
+    Asum *= u_total/Tnorm;
+    rhoBsum /= Tnorm;
+    Tideal = u_total / rhoBsum / Tnorm;
+    Pideal = Tideal * Asum / vfrac_total / u_total;
+  }
+
+  template <typename T>
+  PORTABLE_INLINE_FUNCTION
+  void TryIdealPTE(T *solver) {
+    Real Pideal, Tideal;
+    GetIdealPTE(Pideal, Tideal);
+
+    // temporarily hijack some of the scratch space
+    Real *etemp = jacobian;
+    Real *ptemp = jacobian + nmat;
+    Real *vtemp = jacobian + 2*nmat;
+    Real *rtemp = jacobian + 3*nmat;
+    Real *res = jacobian + 4*nmat;
+    // copy out the initial guess
+    for (int m = 0; m < nmat; ++m) {
+      etemp[m] = sie[m];
+      ptemp[m] = press[m];
+      vtemp[m] = vfrac[m];
+      rtemp[m] = rho[m];
+    }
+    Real res_norm_old = 0.0;
+    for (int m = 0; m < neq; ++m) {
+      res[m] = residual[m];
+      res_norm_old += res[m] * res[m];
```

Comment:
As written the `ResidualNorm` function always computes the norm of the `residual` member variable, so wouldn't immediately apply here.  It would be straightforward to have that function take a `Real *` and then it could work for this too.  On the other hand, this is like one line of code anyway.  Could go either way on this one.